File: sdk/typescript/src/skills.ts
import { AxiosInstance } from "axios";
import { SkillDefinition } from "./types";
import { ENDPOINTS, SKILL_NAMES } from "./constants";

/** Skill registry client with lazy caching and chain expansion. */
export class SkillsApi {
    private readonly http: AxiosInstance;
    private cache: Map<string, SkillDefinition> = new Map();

    constructor(http: AxiosInstance) {
        this.http = http;
    }

    async list(): Promise<SkillDefinition[]> {
        const response = await this.http.get(ENDPOINTS.skills);
        const skills: SkillDefinition[] = response.data;

        for (const skill of skills) {
            this.cache.set(skill.name, skill);
        }

        return skills;
    }

    async get(name: string): Promise<SkillDefinition> {
        const cached = this.cache.get(name);
        if (cached) {
            return cached;
        }

        const response = await this.http.get(`${ENDPOINTS.skills}/${name}`);
        const skill: SkillDefinition = response.data;
        this.cache.set(name, skill);
        return skill;
    }

    async validate(name: string): Promise<boolean> {
        const validNames = Object.values(SKILL_NAMES);
        if (validNames.includes(name as any)) {
            return true;
        }

        try {
            await this.get(name);
            return true;
        } catch {
            return false;
        }
    }

    async getChain(skillName: string): Promise<string[]> {
        if (skillName === SKILL_NAMES.DEEP_AUDIT) {
            return [
                SKILL_NAMES.SECURITY_AUDIT,
                SKILL_NAMES.BEST_PRACTICES,
                SKILL_NAMES.GAS_OPTIMIZATION,
            ];
        }

        return [skillName];
    }

    clearCache(): void {
        this.cache.clear();
    }

    getBuiltinSkills(): typeof SKILL_NAMES {
        return SKILL_NAMES;
    }

    async resolveSkills(skillNames: string[]): Promise<SkillDefinition[]> {
        const resolved: SkillDefinition[] = [];
        const expanded: string[] = [];

        for (const name of skillNames) {
            const chain = await this.getChain(name);
            expanded.push(...chain);
        }

        const unique = [...new Set(expanded)];

        for (const name of unique) {
            const skill = await this.get(name);
            resolved.push(skill);
        }

        return resolved;
    }
}

File: sdk/python/parity_sdk/skills.py
from __future__ import annotations

import httpx

from parity_sdk.types import SkillDefinition, SkillInput, SkillOutput
from parity_sdk.constants import ENDPOINTS, SKILL_NAMES


class SkillsApi:  # Query, cache, and resolve analysis skill definitions
    """Interface for querying and managing analysis skills."""

    def __init__(self, http: httpx.Client) -> None:
        self._http = http
        self._cache: dict[str, SkillDefinition] = {}

    def list(self) -> list[SkillDefinition]:
        """List all available skills."""
        response = self._http.get(ENDPOINTS["skills"])
        response.raise_for_status()
        skills_data = response.json()

        skills: list[SkillDefinition] = []
        for raw in skills_data:
            skill = self._parse_skill(raw)
            self._cache[skill.name] = skill
            skills.append(skill)

        return skills

    def get(self, name: str) -> SkillDefinition:
        """Get a specific skill by name."""
        cached = self._cache.get(name)
        if cached is not None:
            return cached

        response = self._http.get(f"{ENDPOINTS['skills']}/{name}")
        response.raise_for_status()
        skill = self._parse_skill(response.json())
        self._cache[name] = skill
        return skill

    def validate(self, name: str) -> bool:
        """Check if a skill name is valid."""
        valid_names = set(SKILL_NAMES.values())
        if name in valid_names:
            return True

        try:
            self.get(name)
            return True
        except httpx.HTTPStatusError:
            return False

    def get_chain(self, skill_name: str) -> list[str]:
        """Get the execution chain for a skill."""
        if skill_name == SKILL_NAMES["DEEP_AUDIT"]: